homeassistant/components/cloud/backup.py

```diff
@@ -0,0 +1,190 @@
+
+
+class CloudBackupAgent(BackupAgent):
+    """Cloud backup agent."""
+
+    name = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, cloud: Cloud[CloudClient]) -> None:
+        """Initialize the cloud backup sync agent."""
+        super().__init__()
+        self._cloud = cloud
+        self._hass = hass
+        self._backups: list[BaseBackup] = []
+
+    @callback
+    def _get_backup_filename(self) -> str:
+        """Return the backup filename."""
+        return f"{self._cloud.client.prefs.instance_id}.tar"
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        *,
+        path: Path,
+        **kwargs: Any,
+    ) -> None:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :param path: The full file path to download the backup to.
+        """
+        details = await async_files_download_details(
+            self._cloud,
+            storage_type=_STORAGE_BACKUP,
+            filename=self._get_backup_filename(),
+        )
+
+        resp = await self._cloud.websession.get(
+            details["url"],
+            raise_for_status=True,
+        )
+
+        file = await self._hass.async_add_executor_job(path.open, "wb")
+        async for chunk in resp.content.iter_any():
+            await self._hass.async_add_executor_job(file.write, chunk)
+
+    async def async_upload_backup(
+        self,
+        *,
+        path: Path,
+        metadata: BackupUploadMetadata,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param path: The full file path to the backup that should be uploaded.
+        :param metadata: Metadata about the backup that should be uploaded.
+        """
+        if not metadata.protected:
+            raise BackupAgentError("Cloud backups must be protected")
+
+        if not self._cloud.is_logged_in:
+            raise BackupAgentError("Not logged in to cloud")
+
+        def _create_hash_and_get_size() -> tuple[str, int]:
+            """Create file hash and calculate size."""
+            return b64md5(path), path.stat().st_size
+
+        base64md5hash, size = await self._hass.async_add_executor_job(
+            _create_hash_and_get_size
+        )
+
+        details = await async_files_upload_details(
+            self._cloud,
+            storage_type=_STORAGE_BACKUP,
+            filename=self._get_backup_filename(),
+            metadata={
+                "backup_id": metadata.backup_id,
+                "date": metadata.date,
+                "homeassistant_version": metadata.homeassistant,
+                "name": metadata.name,
+                "protected": metadata.protected,
+            },
+            size=size,
+            base64md5hash=base64md5hash,
+        )
+
+        await self._cloud.websession.put(
+            details["url"],
+            data={"file": path.open("rb")},
```

Comment:
How will opening the file work in async context? Is it supported and will the file be streamed?